src/rust/engine/fs/store/src/lib.rs

```diff
@@ -679,13 +679,40 @@ impl Store {
     digest: Digest,
   ) -> BoxFuture<DirectoryMaterializeMetadata, String> {
     let root = Arc::new(Mutex::new(None));
+    let executor = self.local.executor().clone();
     self
       .materialize_directory_helper(
-        destination,
+        destination.clone(),
         RootOrParentMetadataBuilder::Root(root.clone()),
         digest,
       )
-      .map(|()| Arc::try_unwrap(root).unwrap().into_inner().unwrap().build())
+      .and_then(move |()| {
+        let dmm = Arc::try_unwrap(root).unwrap().into_inner().unwrap().build();
+        // We fundamentally materialize files for other processes to read; as such, we must ensure
+        // data is flushed to disk and visible to them as opposed to just our process. Even though
+        // we need to re-open all written files, executing all fsyncs at the end of the
```

Comment:
When I did this in https://github.com/pantsbuild/pants/pull/7887 rather than re-opening the files, I passed around the opened file handles. May or may not be worth doing similarly :)

Reply:
Mm. I'd be concerned about total file handle count for some of our materializes.

Reply:
And jeez: I'd totally forgotten about those patches... must have been in my subconscious somewhere. Thank you!